src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Jeffijoe.MessageFormat.MetadataGenerator.Plural.Parsing
+{
+    public class RuleParser
+    {
+        private string _ruleText;
+        private int _position;
+
+        public RuleParser(string ruleText)
+        {
+            _ruleText = ruleText;
+        }
+
+        public OrCondition[] ParseRuleContent()
+        {
+            var conditions = new List<OrCondition>();
+
+            while (!IsEnd)
+            {
+                if (PeekCurrentChar == '@')
+                {
+                    return conditions.ToArray();
+                }
+
+                var condition = ParseOrCondition();
+                conditions.Add(condition);
+
+                AdvanceWhitespace();
+
+                var character = ConsumeChar();
+                var characterNext = ConsumeChar();
+
+                if (character == '@')
+                {
+                    return conditions.ToArray();
+                }
+                else if (character == 'o' && characterNext == 'r')
+                {
+                    continue;
+                }
+                else
+                {
+                    throw new InvalidCharacterException(character);
+                }
```

Comment:
It's definitely negligible, but we can defer reading the next character in case we have an early return.

```suggestion
                var character = ConsumeChar();
                // This is where the samples start, we don't care about any of those.
                if (character == '@')
                {
                    return conditions.ToArray();
                }

                // We expect the next token to be "or"
                var characterNext = ConsumeChar();
                if (character == 'o' && characterNext == 'r')
                {
                    continue;
                }
                
                throw new InvalidCharacterException(character);
```

Reply:
I like the change! Going with it